Article "Provide in-app purchases with the Digital Goods API" (#3313)
MicrosoftDocs/edge-developer: microsoft-edge/progressive-web-apps-chromium/how-to/digital-goods-api.md

---
title: Get a Microsoft Store billing service by using the Digital Goods API
description: Use the Digital Goods API to get a Microsoft Store billing service.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.service: microsoft-edge
ms.subservice: pwa
ms.date: 11/21/2024
---
# Get a Microsoft Store billing service by using the Digital Goods API

> <!-- todo: improve title (in 3+ spots) -->

Author: Modified.

> <!-- todo: improve the description: field above; sync w/ Sentence 1; develop clear wording for both spots -->

Reviewer: improve the **description:** field above; sync w/ Sentence 1; develop clear wording for both spots
@runyuany

Reviewer: result:
description: How to provide in-app purchases in a Progressive Web App by using the Digital Goods API and the Payment Request API.

If your Progressive Web App (PWA) is listed in the Microsoft Store, you can provide in-app products and subscriptions by using the Digital Goods API and the Payment Request API.

See also:
* [Digital Goods API For Microsoft Store PWA Explainer](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/PwaDigitalGoods/explainer.md)
* [Payment Request API](https://www.w3.org/TR/payment-request/) at W3C.

<!-- ====================================================================== -->
## Digital Goods API

The Digital Goods API is an interface between your PWA app and Microsoft Store.  The Digital Goods API supports:
* Querying the details of a digital item from the Microsoft Store backend, such as the item's name, description, and regional price.
* Consuming or acknowledging purchases.
* Checking the digital items that are currently owned by the user.
* Checking the purchase history of the user.

> Before making the Digital Goods API available in your app, make sure that - _todo: finish sentence_

Reviewer: finish sentence "Before making the Digital Goods API available in your app, make sure that"
@runyuany

---
title: Provide in-app purchases with the Digital Goods API
description: How to provide in-app purchases in a Progressive Web App by using the Digital Goods API and the Payment Request API.
author: MSEdgeTeam
ms.author: msedgedevrel
ms.topic: conceptual
ms.service: microsoft-edge
ms.subservice: pwa
ms.date: 11/25/2024
---
# Provide in-app purchases with the Digital Goods API

If your Progressive Web App (PWA) is listed in the Microsoft Store, you can provide in-app products and subscriptions by using the Digital Goods API and the Payment Request API.

<!-- ====================================================================== -->
## Digital Goods API

> The Digital Goods API is an interface between your PWA app and Microsoft Store.  The Digital Goods API supports:

Reviewer 2: ```suggestion
The Digital Goods API is an interface between your PWA app and the Microsoft Store.  The Digital Goods API supports:
```

* Querying the details of a digital item from the Microsoft Store backend, such as the item's name, description, and regional price.
* Consuming or acknowledging purchases.
* Checking the digital items that are currently owned by the user.
* Checking the purchase history of the user.

See:
* [Digital Goods API For Microsoft Store PWA Explainer](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/PwaDigitalGoods/explainer.md)
* [Digital Goods API: Draft Community Group Report](https://wicg.github.io/digital-goods/)

<!-- ====================================================================== -->
## Payment Request API

> The Payment Request API⁠⁠ handles the actual payment transaction when a purchase is made by user.  The Payment Request API uses the item details that the Digital Goods API provides, to make the in-app purchase by using whichever billing payment method the user has set up at the Microsoft Store.

Reviewer 2: ```suggestion
The Payment Request API⁠⁠ handles the actual payment transaction when a purchase is made by a user.  The Payment Request API uses the item details that the Digital Goods API provides, to make the in-app purchase by using whichever billing payment method the user has set up at the Microsoft Store.
```

> See [Payment Request API](https://www.w3.org/TR/payment-request/) at W3C.

Reviewer 2: ```suggestion
See [Payment Request API](https://developer.mozilla.org/docs/Web/API/Payment_Request_API) at MDN.
```

Let's link to MDN, which is much more developer friendly than the W3C spec.

<!-- ====================================================================== -->
## Checking whether the Digital Goods API is available

> To detect whether you've correctly enabled the API on your website, check for the `getDigitalGoodsService` method in the window object:

Reviewer 2: This is phrased as if the developer needs to take action to enable the API, but we're not telling them what action that is. Is there an origin trial? Is there a flag to enable somewhere? Does this come with a pre-release version of Edge?

Reviewer: now doc'd:
https://review.learn.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/how-to/digital-goods-api?branch=pr-en-us-3324#enable-the-digital-goods-api
@captainbrosset

```javascript
if ('getDigitalGoodsService' in window) {
  // Digital Goods API is supported!
} else {
  console.log('DigitalGoodsService is not available.');
  // Use other payment method
}
```

<!-- ====================================================================== -->
## Connecting to the Microsoft Store Billing service (`getDigitalGoodsService` method)

Use the `getDigitalGoodsService` method to connect to the Microsoft Store Billing service.

The Digital Goods API was designed to be compatible with various browsers and digital stores, similar to how the Payment Request API is browser-agnostic and can be used with different payment providers.  To retrieve an instance of the service for Microsoft Store Billing, pass the string `"https://store.microsoft.com/billing"` as the payment method to the `getDigitalGoodsService` method.

> If the method throws an error, the Microsoft Store Billing payment method is not available (such as when the user is accessing your PWA through the browser).  Alternatively, consider providing a different payment method for transactions.

Reviewer 2: I think we should expand on the reasons that might make the payment method not available. When is the method available exactly? What are the requirements for it to be available? PWA installed from the store? Any other required settings?

Reviewer: doc'd:
https://review.learn.microsoft.com/en-us/microsoft-edge/progressive-web-apps-chromium/how-to/digital-goods-api?branch=pr-en-us-3324#connecting-to-the-microsoft-store-billing-service-windowgetdigitalgoodsservice-method
@captainbrosset

```javascript
if (window.getDigitalGoodsService === undefined) {
  // Digital Goods API is not supported in this context.
  return;
}
try {
  const digitalGoodsService = await window.getDigitalGoodsService("https://store.microsoft.com/billing");
  // Use the service here.
  ...
} catch (error) {
  // Our preferred service provider is not available.

>   // Use a normal web-based payment flow.

Reviewer 2: ```suggestion
  // Use a web-based payment flow instead.
```

I don't think "normal" is warranted here, it implies the digitial goods API flow is not normal.

  return;
}
```

<!-- ====================================================================== -->
## Querying item details (`getDetails` method)

Use the `getDetails` method to query item details.

After connecting the Digital Goods service to Microsoft Store, you can use the API to access product and purchase information.  The `getDetails` method lets you get information about the items you’ve set up in the Partner Center.  Display information such as the product title, description, and price in your app UI, so the user knows what's available for purchase.

The `getDetails` method takes a list of item IDs, which correspond to the product IDs of the in-app products and subscriptions you created in the Partner Center.

```javascript

> details = await digitalGoodsService.getDetails(['shiny_sword', 'gem', 'monthly_subscription']);
> for (item of details) {
>   const priceStr = new Intl.NumberFormat(
>       locale,
>       {style: 'currency', currency: item.price.currency}
>     ).format(item.price.value);
>   AddShopMenuItem(item.itemId, item.title, priceStr, item.description);
> }

Reviewer 2: ```suggestion
const itemDetails = await digitalGoodsService.getDetails(['shiny_sword', 'gem', 'monthly_subscription']);

for (item of itemDetails) {
  const priceStr = new Intl.NumberFormat(
      locale,
      {style: 'currency', currency: item.price.currency}
    ).format(item.price.value);

  // Do something with the item's data, such as displaying it in the PWA's UI.
  displayProductItem(item.itemId, item.title, priceStr, item.description);
}
```

Changing `details` to `itemDetails` since that's what's used in the next paragraph.

```

The returned `itemDetails` sequence may be in any order, and might not include an item if the item doesn't exist on the server (that is, if there's not a 1:1 correspondence between the input list and output list).

> The item ID is a string that represents the primary key of the items.  As in Microsoft Store, the item ID is `InAppOfferToken`.  There is no function to get a list of item IDs; item IDs should be hardcoded in the client code or fetched from your own server (the developer's server).

Reviewer 2: ```suggestion
The item ID is a string that represents the primary key of the items.  In the Microsoft Store, the item ID is `InAppOfferToken`.  There is no function to get a list of item IDs; item IDs should be hardcoded in the client code or fetched from your own server (the developer's server).
```

The item's `price` is a `PaymentCurrencyAmount` that contains the current price of the item in the user's current region and currency.  The `price` is designed to be formatted for the user's current locale by using `Intl.NumberFormat`, as shown above.

See also:
* [StoreProduct.InAppOfferToken Property](/uwp/api/windows.services.store.storeproduct.inappoffertoken)
* [PaymentCurrencyAmount dictionary](https://www.w3.org/TR/payment-request/#dom-paymentcurrencyamount) in _Payment Request API_ at W3C.
* [Intl.NumberFormat](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat) at MDN.

<!-- ====================================================================== -->
## Purchasing an item (`show` method)

> Use the `show` method to purchase an item, after you construct a request that contains item details.

Reviewer 2: ```suggestion
Use the `show` method to purchase an item, after you construct a request that contains the item details.
```

Once your products and details are displayed to the user, you can implement the purchase flow by using the Payment Request API.  When combined with the Digital Goods API, the only required input parameter is `methodData`.

Use the `supportedMethods` member of the `methodData`⁠⁠ parameter in the `PaymentRequest` to identify Microsoft Store Billing as the payment method with the string `"https://store.microsoft.com/billing"`.  Then in the `data` member, pass along the item ID as the `sku`.

```javascript

> const details = await digitalGoodsService.getDetails(['monthly_subscription']);
> const item = details[0];
> const request = new PaymentRequest(
>   [{supportedMethods: 'https://store.microsoft.com/billing',
>     data: {sku: item.itemId}}
>   ]
> );

Reviewer 2: ```suggestion
const details = await digitalGoodsService.getDetails(['monthly_subscription']);
const item = details[0];

const request = new PaymentRequest([
  {
    supportedMethods: 'https://store.microsoft.com/billing',
    data: { sku: item.itemId }
  }
]);
```

To mark a purchase as "consumed", use the `consume` method:

```javascript
digitalGoodsService.consume(purchaseToken);
```

<!-- ====================================================================== -->
## Checking existing purchases (`listPurchases` method)

Use the `listPurchases` method to check existing purchases.  This method returns information about the user's existing purchases.  This method allows a client to get a list of items that are currently owned or purchased by the user.  This may be necessary, to do either of the following:

* Check for entitlements, such as whether a subscription, promotional code, or permanent upgrade is active.

* Recover from network interruptions during a purchase, such as when the item is purchased but not yet acknowledged.

The `listPurchases` method returns item IDs and purchase tokens.  Before you grant an entitlement, you should verify the returned item ID or the returned purchase token, by using a direct developer-to-provider API, as shown below:

```javascript

> purchases = await digitalGoodsService.listPurchases();
> for (p of purchases) {
>   VerifyAndGrantEntitlement(p.itemId, p.purchaseToken);
> }

Reviewer 2: ```suggestion
const purchaseList = await digitalGoodsService.listPurchases();

for (const purchase of purchaseList) {
  // Handle the purchase data in your PWA.
  verifyAndGrantEntitlement(purchase.itemId, purchase.purchaseToken);
}
```

```

The `listPurchases` method doesn't return consumed products or expired subscriptions.

<!-- ====================================================================== -->
## Getting the purchase history (`listPurchaseHistory` method)

Use the `listPurchaseHistory` method to get the purchase history.  This method returns a list that shows the most recent purchase made by the user for each item, regardless of whether the purchase is expired, canceled, or consumed.  This method returns a list of `PurchaseDetails` containing the `itemId` and `purchaseToken` for each purchase.

```javascript

> purchases = await digitalGoodsService.listPurchaseHistory();
> for (p of purchases) {
>   VerifyAndCheckExpiredEntitlement(p.itemId, p.purchaseToken);
> }

Reviewer 2: ```suggestion
const purchaseList = await digitalGoodsService.listPurchaseHistory();

for (const purchase of purchaseList) {
  // Handle the expired purchase data in your PWA.
  verifyAndCheckExpiredEntitlement(purchase.itemId, purchase.purchaseToken);
}
```

```javascript
purchases = await digitalGoodsService.listPurchaseHistory();
for (p of purchases) {
  VerifyAndCheckExpiredEntitlement(p.itemId, p.purchaseToken);
}
```

<!-- ====================================================================== -->
## See also

* [Digital Goods API For Microsoft Store PWA Explainer](https://github.com/MicrosoftEdge/MSEdgeExplainers/blob/main/PwaDigitalGoods/explainer.md)
* [StoreProduct.InAppOfferToken Property](/uwp/api/windows.services.store.storeproduct.inappoffertoken)

W3C:
* [Digital Goods API: Draft Community Group Report](https://wicg.github.io/digital-goods/)
* [Payment Request API](https://www.w3.org/TR/payment-request/)
* [5. PaymentCurrencyAmount dictionary](https://www.w3.org/TR/payment-request/#dom-paymentcurrencyamount) in _Payment Request API_.

MDN:
* [Intl.NumberFormat](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat)

> * [display_override](https://developer.mozilla.org/docs/Web/Manifest/display_override)<!-- todo: delete link? "override" not in article body   -->

Reviewer: delete "display_override" Ref link?  "override" not in article body
https://developer.mozilla.org/docs/Web/Manifest/display_override
@runyuany

  ...
} catch (error) {
  // Our preferred service provider is not available.
  // Use a normal web-based payment flow.
  return;
}
```

<!-- ====================================================================== -->
## Querying item details

Use the `getDetails` method to query item details.

After connecting the Digital Goods service to Microsoft Store, you can use the API to access product and purchase information.  The `getDetails` method lets you get information about the items you’ve set up in the Partner Center.  Display information such as the product title, description, and price in your app UI, so the user knows what's available for purchase.

The `getDetails` method takes a list of item IDs, which correspond to the product IDs of the in-app products and subscriptions you created in the Partner Center.

```javascript
details = await digitalGoodsService.getDetails(['shiny_sword', 'gem', 'monthly_subscription']);
for (item of details) {
  const priceStr = new Intl.NumberFormat(
      locale,
      {style: 'currency', currency: item.price.currency}
    ).format(item.price.value);
  AddShopMenuItem(item.itemId, item.title, priceStr, item.description);
}
```

> The returned `itemDetails` sequence may be in any order, and may not<!-- todo: might not? --> include an item if the item doesn't exist on the server (that is, if there's not a 1:1 correspondence between the input list and output).<!-- todo: output list? -->

Reviewer: ```suggestion
The returned `itemDetails` sequence may be in any order, and might not include an item if the item doesn't exist on the server (that is, if there's not a 1:1 correspondence between the input list and output list).
```

Author: Modified, thanks.
